service/history/workflow/update/update.go

```diff
@@ -26,85 +26,392 @@ package update
 
 import (
 	"context"
+	"sync/atomic"
 
-	"github.com/pborman/uuid"
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
 	failurepb "go.temporal.io/api/failure/v1"
+	historypb "go.temporal.io/api/history/v1"
+	protocolpb "go.temporal.io/api/protocol/v1"
 	updatepb "go.temporal.io/api/update/v1"
 
 	"go.temporal.io/server/common/future"
+	"go.temporal.io/server/internal/effect"
 )
 
 type (
+	// EventStore is the interface that an Update needs to read and write events
+	// and to be notified when buffered writes have been flushed. It is the
+	// expectation of this code that writes to EventStore will return before the
+	// data has been made durable. Callbacks attached to the EventStore via
+	// OnAfterCommit and OnAfterRollback *must* be called after the EventStore
+	// state is successfully written or is discarded.
+	EventStore interface {
+		effect.Controller
+
+		// AddWorkflowExecutionUpdateAcceptedEvent writes an update accepted
+		// event. The data may not be durable when this function returns.
+		AddWorkflowExecutionUpdateAcceptedEvent(
+			updateID string,
+			accpt *updatepb.Acceptance,
+		) (*historypb.HistoryEvent, error)
+
+		// AddWorkflowExecutionUpdateCompletedEvent writes an update completed
+		// event. The data may not be durable when this function returns.
+		AddWorkflowExecutionUpdateCompletedEvent(
+			resp *updatepb.Response,
+		) (*historypb.HistoryEvent, error)
+	}
+
+	// Update is a state machine for the update protocol. It reads and writes
+	// messages from the go.temporal.io/api/update/v1 package. The update state
+	// machine is straightforward except in that it provides "provisional"
+	// in-between states where the update has received a message that has
+	// updated its internal state but those updates have not been made visible
+	// to clients yet (e.g. accepted or outcome futures have not been set yet).
+	// The effects are bound to the EventStore's effect.Set and will be
+	// triggered withen those effects are applied.
 	Update struct {
-		state                 state
-		request               *updatepb.Request
-		afterBufferedEventNum int
-		messageID             string
-		protocolInstanceID    string
-		outcome               *future.FutureImpl[*updatepb.Outcome]
-		pendingOutcome        *updatepb.Outcome
+		// accessed only while holding workflow lock
+		id              string
```

Comment:
`id` field in `Update` struct reads as `UpdateID`. I believe `MessageID` is a better name here.

Reply:
This is the UpdateID though. MessageID is built just-in-time when sending the protocolpb.Message